src/LocalDate.php

```diff
@@ -759,9 +762,12 @@ public function jsonSerialize(): string
      */
     public function toISOString(): string
     {
-        $pattern = ($this->year < 0 ? '%05d' : '%04d') . '-%02d-%02d';
-
-        return sprintf($pattern, $this->year, $this->month, $this->day);
+        // This code is optimized for high performance
+        return ($this->year < 1000 ? ($this->year < 0 ? '-' : '') . str_pad((string) abs($this->year), 4, '0', STR_PAD_LEFT) : $this->year)
+            . '-'
+            . ($this->month < 10 ? '0' . $this->month : $this->month)
+            . '-'
+            . ($this->day < 10 ? '0' . $this->day : $this->day);
```

Comment:
I think this can be slightly optimized further:

```suggestion
        return (
            ($this->year < 1000 && $this->year > -1000)
                ? ($this->year < 0
                    ? '-' . str_pad((string) -$this->year, 4, '0', STR_PAD_LEFT)
                    : str_pad((string) $this->year, 4, '0', STR_PAD_LEFT)
                )
                : $this->year
            )
            . '-'
            . ($this->month < 10 ? '0' . $this->month : $this->month)
            . '-'
            . ($this->day < 10 ? '0' . $this->day : $this->day);
```

Reply:
I don't think I agree with this change ; on the upside, it removes the call to `abs()` (which I've not noticed to be a performance bottleneck) inside an edge-case branch, but on the downside it adds a second condition `&& $this->year > -1000` that will be executed for every call to the method (and it's more verbose/hard to follow IMHO, with yet one more nested ternary).

As 99.99% of dates used in today's web applications will be between 1950 and now+50 years, that change goes against the idea of "optimizing for the most common path and putting more expensive code behind edge-case conditions".

Reply:
While I sympathize with the readability argument, I think we've already killed readability, so why not push it to the limits?

> but on the downside it adds a second condition && $this->year > -1000 that will be executed for every call to the method.

No, the second condition [won't be executed](https://3v4l.org/E4T1M) if the first condition is `false`, so it won't add any overhead for the common case.

@tigitz Do you have an opinion on this?

Reply:
Oh right, my bad! It's an "and" and not an "or"... Okay, I'll apply this change then.

For the readability argument, I liked that it was still fitting one line per part (year, month, etc). With this one I'll try to find how to format it best.